driver-core/src/main/com/mongodb/connection/ElementExtendingBsonWriter.java

```diff
@@ -26,17 +27,25 @@
 class ElementExtendingBsonWriter extends LevelCountingBsonWriter {
     private final List<BsonElement> extraElements;
 
-    ElementExtendingBsonWriter(final BsonWriter writer, final List<BsonElement> extraElements) {
+    ElementExtendingBsonWriter(final BsonBinaryWriter writer, final List<BsonElement> extraElements) {
         super(writer);
         this.extraElements = extraElements;
     }
 
     @Override
     public void writeEndDocument() {
         if (getCurrentLevel() == 0) {
-            writeElements(getBsonWriter(), extraElements);
+            writeElements(getBsonBinaryWriter(), extraElements);
         }
         super.writeEndDocument();
     }
 
+    @Override
+    public void pipe(final BsonReader reader) {
+        if (getCurrentLevel() == -1) {
```

Comment:
There was actually a bug in the original version, which compared the current level to 0 instead of -1.  It wasn't caught then either because of the test bug.